internal/future/serving/tls_server.go

```diff
@@ -0,0 +1,225 @@
+	httpServeWaiter chan error
+	httpMux         *http.ServeMux
+	proxyMux        *runtime.ServeMux
+	httpServer      *http.Server
+}
+
+func (s *tlsServer) start(params *Params) (func(), error) {
+	s.grpcServeWaiter = make(chan error)
+	s.httpServeWaiter = make(chan error)
+	var serverStartWaiter sync.WaitGroup
+
+	grpcAddress := fmt.Sprintf("localhost:%d", s.grpcLh.Number())
+
+	grpcListener, err := s.grpcLh.Obtain()
+	if err != nil {
+		return func() {}, errors.WithStack(err)
+	}
+	s.grpcListener = grpcListener
+
+	rootCaCert, err := TrustedCertificates(params.rootCaPublicCertificateFileData)
+	if err != nil {
+		return func() {}, errors.WithStack(err)
+	}
+	credsForProxyToGrpc, err := ClientCredentialsFromFileData(params.publicCertificateFileData, "")
+	if err != nil {
+		return func() {}, errors.WithStack(err)
+	}
+
+	grpcTLSCertificate, err := CertificateFromFileData(params.publicCertificateFileData, params.privateKeyFileData)
+	if err != nil {
+		return func() {}, errors.WithStack(err)
+	}
+	creds := credentials.NewServerTLSFromCert(grpcTLSCertificate)
+	s.grpcServer = grpc.NewServer(grpc.Creds(creds), grpc.StatsHandler(&ocgrpc.ServerHandler{}))
+	// Bind gRPC handlers
+	for _, handlerFunc := range params.handlersForGrpc {
+		handlerFunc(s.grpcServer)
+	}
+
+	serverStartWaiter.Add(1)
+	go func() {
+		serverStartWaiter.Done()
+		s.grpcServeWaiter <- s.grpcServer.Serve(s.grpcListener)
+	}()
+
+	// Start HTTP server
+	httpListener, err := s.httpLh.Obtain()
+	if err != nil {
+		return func() {}, errors.WithStack(err)
+	}
+	s.httpListener = httpListener
+	// Bind gRPC handlers
+	ctx := context.Background()
+
+	httpsToGrpcProxyOptions := []grpc.DialOption{grpc.WithTransportCredentials(credsForProxyToGrpc)}
+
+	for _, handlerFunc := range params.handlersForGrpcProxy {
+		if err = handlerFunc(ctx, s.proxyMux, grpcAddress, httpsToGrpcProxyOptions); err != nil {
+			return func() {}, errors.WithStack(err)
+		}
+	}
+
+	// Bind HTTPS handlers
+	s.httpMux.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "ok")
+	})
+	s.httpMux.Handle("/", s.proxyMux)
+	zpages.Handle(s.httpMux, "/debug")
+	s.httpServer = &http.Server{
+		Addr:    s.httpListener.Addr().String(),
+		Handler: s.httpMux,
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{*grpcTLSCertificate},
+			ClientCAs:    rootCaCert,
+			ClientAuth:   tls.RequireAndVerifyClientCert,
+			NextProtos:   []string{http2WithTLSVersionID}, // https://github.com/grpc-ecosystem/grpc-gateway/issues/220
+		},
+	}
+	serverStartWaiter.Add(1)
+	go func() {
+		serverStartWaiter.Done()
+		tlsListener := tls.NewListener(s.httpListener, s.httpServer.TLSConfig)
+		s.httpServeWaiter <- s.httpServer.Serve(tlsListener)
+	}()
+
+	// Wait for the servers to come up.
+	return serverStartWaiter.Wait, nil
+}
+
+func (s *tlsServer) stop() {
+	s.grpcServer.Stop()
+	// Close the TCP listener, it'll always report that the port is closed so just log the error.
```

Comment:
nit - 3 identical repeated comments - just aggregate to 1 before the code block?

Reply:
No idea how this comment keeps on coming back.